File: sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/SpawnedThreads.java
package io.littlehorse.sdk.wfsdk;

import io.littlehorse.sdk.common.proto.WaitForThreadsNode;
import io.littlehorse.sdk.common.proto.WaitForThreadsStrategy;
import io.littlehorse.sdk.wfsdk.internal.FixedSpawnedThreads;

/**
 * Collection-like handle for a set of spawned threads.
 */
public interface SpawnedThreads {

    /**
     * Builds the wait node for this spawned thread collection.
     *
     * @param strategy waiting strategy to apply
     * @return wait node definition
     */
    WaitForThreadsNode buildNode(WaitForThreadsStrategy strategy);

    /**
     * Creates a SpawnedThreads instance from explicit thread handles.
     *
     * @param threads spawned thread handles
     * @return spawned thread collection
     */
    static SpawnedThreads of(SpawnedThread... threads) {
        return new FixedSpawnedThreads(threads);
    }
}

File: sdk-dotnet/LittleHorse.Sdk/Workflow/Spec/SpawnedThreads.cs
using LittleHorse.Sdk.Common.Proto;

namespace LittleHorse.Sdk.Workflow.Spec;

/// <summary>
/// Represents a collection of spawned threads that are fixed at the time of creation.
/// </summary>
public interface SpawnedThreads
{
    /// <summary>
    /// Builds a <see cref="WaitForThreadsNode"/> that represents the spawned threads.
    /// </summary>
    /// <returns>WaitForThreadsNode</returns>
    WaitForThreadsNode BuildNode();

    /// <summary>
    /// Creates a new instance of <see cref="SpawnedThreads"/> with the specified spawned threads.
    /// </summary>
    /// <param name="threads">Any number spawned threads.</param>
    /// <returns></returns>
    static SpawnedThreads Of(params SpawnedThread[] threads) 
    {
        return new FixedSpawnedThreads(threads);
    }
}